File: Easy/20/20.cpp
#include <iostream>
#include <stack>
#include <unordered_map>
#include <unordered_set>

using namespace std;

unordered_set<char> openBrackets = {'(', '[', '{'};
unordered_map<char, char> closeToOpen = {{')', '('}, {']', '['}, {'}', '{'}};

class Solution {
public:
  bool isValid(string s) {
    // Last open bracket must be the first to be closed (LIFO)
    stack<char> history;
    for (char c : s) {
      if (openBrackets.count(c))
        history.push(c);
      else {
        if (history.empty() || history.top() != closeToOpen[c])
          return false;
        history.pop();
      }
    }
    return history.empty() ? true : false;
  }
};

int main() {
  cout << "It's Showtime Folks!" << endl;
  Solution s;
  cout << "Solution: " << s.isValid("()") << endl;
  cout << "Solution: " << s.isValid("()[]{}") << endl;
  cout << "Solution: " << s.isValid("(]") << endl;
  cout << "Solution: " << s.isValid("([])") << endl;
  cout << "Solution: " << s.isValid("([)]") << endl;
  return 0;
}

File: Easy/20/20.ts
export {};

const closeToOpen = new Map([
  [")", "("],
  ["]", "["],
  ["}", "{"],
]);

function isValid(s: string): boolean {
  let history: string[] = [];
  for (const c of s) {
    if (closeToOpen.has(c)) {
      if (
        history.length === 0 ||
        history[history.length - 1] !== closeToOpen.get(c)
      )
        return false;
      history.pop();
    } else history.push(c);
  }
  return history.length === 0 ? true : false;
}

console.log("It's Showtime Folks!");
console.log(`Solution: ${isValid("()")}`);
console.log(`Solution: ${isValid("()[]{}")}`);
console.log(`Solution: ${isValid("(]")}`);
console.log(`Solution: ${isValid("([])")}`);
console.log(`Solution: ${isValid("([)]")}`);
